### ISSUE TITLE
Moving a class to an upstream SourceSet causes Java incremental compilation to fail

### ISSUE THREAD

**Issue by isker** [gradle/gradle#9380]
TL;DR is in the README of https://github.com/isker/moving-between-sourcesets-bug

If you have a `main` SourceSet with 2 directories each containing one class, compile it, then move one of those directories to a new `upstream` SourceSet (having `main` depend on the output of `upstream`), incremental compilation for `main` fails as Gradle tries to clean the outputs of the previous compilation and sees that the class does not belong to `main`.

### Expected Behavior
Build doesn't fail.  Either incremental compilation needs to not throw this exception or we need to detect the situation and do full recompilation.

### Current Behavior
Build failure (full stacktrace in the linked README):
```
Caused by: java.lang.IllegalArgumentException: Unable to find source java class: '/private/tmp/moving-between-sourcesets-bug/src/upstream/java/other/sourceset/dude/Thing.java' because it does not belong to any of the source dirs: '[/private/tmp/moving-between-sourcesets-bug/src/main/java]'
        at org.gradle.api.internal.tasks.compile.incremental.recomp.SourceToNameConverter.getClassName(SourceToNameConverter.java:44)
        at org.gradle.api.internal.tasks.compile.incremental.recomp.JavaChangeProcessor.processChange(JavaChangeProcessor.java:34)
        at org.gradle.api.internal.tasks.compile.incremental.recomp.InputChangeAction.execute(InputChangeAction.java:47)
        at org.gradle.api.internal.tasks.compile.incremental.recomp.InputChangeAction.execute(InputChangeAction.java:26)
        at org.gradle.api.internal.changedetection.changes.ChangesOnlyIncrementalTaskInputs.doRemoved(ChangesOnlyIncrementalTaskInputs.java:51)
        at org.gradle.api.internal.changedetection.changes.StatefulIncrementalTaskInputs.removed(StatefulIncrementalTaskInputs.java:46)
        at org.gradle.api.internal.tasks.compile.incremental.recomp.CurrentCompilation.visitChanges(CurrentCompilation.java:52)
        at org.gradle.api.internal.tasks.compile.incremental.recomp.RecompilationSpecProvider.processOtherChanges(RecompilationSpecProvider.java:86)
        at org.gradle.api.internal.tasks.compile.incremental.recomp.RecompilationSpecProvider.provideRecompilationSpec(RecompilationSpecProvider.java:41)
        at org.gradle.api.internal.tasks.compile.incremental.SelectiveCompiler.execute(SelectiveCompiler.java:63)
        at org.gradle.api.internal.tasks.compile.incremental.SelectiveCompiler.execute(SelectiveCompiler.java:37)
        at org.gradle.api.internal.tasks.compile.incremental.IncrementalResultStoringCompiler.execute(IncrementalResultStoringCompiler.java:60)
        at org.gradle.api.internal.tasks.compile.incremental.IncrementalResultStoringCompiler.execute(IncrementalResultStoringCompiler.java:44)
        at org.gradle.api.internal.tasks.compile.CompileJavaBuildOperationReportingCompiler$2.call(CompileJavaBuildOperationReportingCompiler.java:59)
        at org.gradle.api.internal.tasks.compile.CompileJavaBuildOperationReportingCompiler$2.call(CompileJavaBuildOperationReportingCompiler.java:51)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.compile.CompileJavaBuildOperationReportingCompiler.execute(CompileJavaBuildOperationReportingCompiler.java:51)
        at org.gradle.api.tasks.compile.JavaCompile.performCompilation(JavaCompile.java:154)
        at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:122)
```
The failure can be resolved only by cleaning.  Repeating the compileJava invocation fails the same way!

### Context
The real-world use case was moving verbose generated code out of `main` to a new upstream SourceSet that could be compiled with far fewer dependencies than the versioned code, cached separately, etc.

### Steps to Reproduce 
See README.txt at https://github.com/isker/moving-between-sourcesets-bug.  This is the specific transformation that breaks incremental compilation https://github.com/isker/moving-between-sourcesets-bug/compare/master..bug.

### Your Environment
Gradle 5.4.1

**Comment by blindpirate** [comment]
Closed by https://github.com/gradle/gradle/pull/9848
